### PR TITLE
WIP: Keep RPMs as artifacts

### PR DESCRIPTION
**Pull request by theobarberbany**

### CODE REVIEW

**Comment by theobarberbany** [comment]
/pj-rehearse pull-ci-openshift-cloud-provider-aws-master-images

**Comment by theobarberbany** [comment]
/pj-rehearse pull-ci-openshift-cloud-provider-aws-master-images

**Comment by theobarberbany** [comment]
/pj-rehearse pull-ci-openshift-cloud-provider-aws-master-images

**Comment by theobarberbany** [comment]
/pj-rehearse pull-ci-openshift-cloud-provider-aws-master-images

File: ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml [hunk 1]
```diff
@@ -61,6 +61,22 @@ promotion:
...
+      as:
+        - rpms
+  to: artifacts
+- output_rpm_to_artifacts_step:
+    from: artifacts
+    optional: true
+    to:
+      name: stable
+      namespace: ""
+      tag: artifacts
```
Comment (LorbusChris) [inline ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml:79]: This isn't going to work I believe. You'll need to create an intermediate image here (e.g. cloud-provider-aws-rpms), promote that, and then import that in the job that owns and creates the artifacts image (https://github.com/openshift/release/blob/master/ci-operator/config/openshift/kubernetes/openshift-kubernetes-master.yaml#L140-L142), and also adapt the Dockerfile literal used there (https://github.com/openshift/release/blob/master/ci-operator/config/openshift/kubernetes/openshift-kubernetes-master.yaml#L128)

Comment (theobarberbany) [reply ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml:79]: I thought the same when I started this work, but think what's above *might* ? work and be more simple. 

I'll try to explain my thought process:

`openshift/kubernetes` has the following steps: 

`pipeline_image_cache_step`: 

We `echo` a dockerfile to `_output/local/releases/Dockerfile` and then build go binaries:  `openshift-hack/build-go.sh`;.


Expanding the dockerfile to make it readable: 

```bash
      mkdir -p _output/local/releases;
      touch _output/local/releases/CHECKSUM; 
      echo 
      $'
      FROM bin AS bin
      FROM rpms AS rpms
      FROM centos:7 
      COPY --from=bin /go/src/k8s.io/kubernetes/_output/local/releases /srv/zips/ 
      COPY --from=oc-rpms /go/src/github.com/openshift/oc/_rpmbuild/SOURCES/* /srv/zips/ 
      COPY --from=rpms /go/src/k8s.io/kubernetes/_output/local/releases/rpms/* /srv/repo/ 
      COPY --from=oc-rpms /go/src/github.com/openshift/oc/_rpmbuild/RPMS/x86_64/* /srv/repo/
      '
      > _output/local/releases/Dockerfile; openshift-hack/build-go.sh;
```

This does some things we don't need:
- zipped binaries (or any binaries really here)
- anything from the build `SOURCES` directory (its untarring the source directory built from git)
- the `oc` binaries (hence nothing from `oc-rpms`?  (I'm unsure about this one). I think this is due to the complexity of `o/k` and shipping additional binaries. 

`project_directory_image_build_step`:

We build the dockerfile outlined above into `artifacts`. 

`output_image_tag_step:`

We promote our artifacts intermediary image to `stable:artifacts`.

We already have `base_machine_with_rpms` above, being used by the install/upgrade job that does a `COPY --from=build /*.rpm /` https://github.com/openshift/release/blob/ec057bb2714501e68d1f6d57d90d724d8f2cbeb9/ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml#L48 and that uses the build location defined here: https://github.com/openshift/release/blob/ec057bb2714501e68d1f6d57d90d724d8f2cbeb9/ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml#L85

My thinking is we can use the `dockerfile_literal` to create an intermediary image that copies the rpms to`/srv/rpms` where they're expected by the `rpm-artifacts-copy` skipping the equivalent of the `pipeline_image_cache_step`. We already have `base_machine_with_rpms` as well, which we have a known rpm location for so we don't need to have a step like the creation of `oc-rpms`?

Comment (theobarberbany) [reply ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml:79]: Ah I'm still not entirely following. I think what you're saying is that the `artifacts` image can only be updated by the `openshift-kubernetes-master` job?

Comment (theobarberbany) [reply ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml:79]: Would it not make sense to trigger https://github.com/okd-project/okd-coreos-pipeline/blob/main/base/tekton.dev/tasks/rpm-artifacts-copy.yaml#L20 on a different image, specifically for these images and leave the `o/k` job alone?

Comment (theobarberbany) [reply ci-operator/config/openshift/cloud-provider-aws/openshift-cloud-provider-aws-master.yaml:79]: Ok - https://github.com/openshift/release/pull/48208 makes more sense to me! Let's close in favour of it.

**Pull request closed without merging**